app/upload/views.py

```diff
@@ -67,81 +66,87 @@ def post(request_id):
     file_ = files[next(files.keys())]
     filename = secure_filename(file_.filename)
     is_update = eval_request_bool(request.form.get('update'))
-    response_id = request.form.get('response_id') if is_update else None
-    if upload_exists(request_id, filename, response_id):
-        response = {
-            "files": [{
-                "name": filename,
-                "error": "A file with this name has already "
-                         "been uploaded for this request."
-                # TODO: "link": <link-to-existsing-file> ? would be nice
-            }]
-        }
-    else:
-        upload_path = os.path.join(
-            current_app.config['UPLOAD_QUARANTINE_DIRECTORY'],
-            request_id)
-        if not os.path.exists(upload_path):
-            os.mkdir(upload_path)
-        filepath = os.path.join(upload_path, filename)
-        key = get_upload_key(request_id, filename, is_update)
+    post_upload = False
+    if is_update and has_permission(permission.EDIT_FILE):
+        post_upload = True
+    elif has_permission(permission.ADD_FILE):
+        post_upload = True
+    if post_upload:
```

Comment:
This doesn't take into account a user that is attempting to edit a file but only has ADD_FILE permissions. If the `is_update` check fails because a user cannot EDIT_FILE, the presence of ADD_FILE would still allow the upload.

```python
if (is_update and has_permission(permission.EDIT_FILE) or (not is_update and has_permission(permission.ADD_FILE):
```
should work... and it's 1 readable line instead of 6.

**Something like this should also be done for DELETE upload/ (route directly below this one).**